pystiche_papers/sanakoyeu_et_al_2018/_loss.py

```diff
@@ -1,18 +1,261 @@
-from typing import Optional
+from abc import abstractmethod
+from collections import OrderedDict
+from typing import Iterator, Optional, Sequence, Union, cast
 
-from pystiche.enc.encoder import SequentialEncoder
-from pystiche.ops.comparison import FeatureReconstructionOperator
+import torch
+import torch.nn as nn
+from torch.nn.functional import binary_cross_entropy_with_logits
 
-from ._modules import TransformerBlock
+from pystiche import enc, ops
+from pystiche.enc import Encoder, MultiLayerEncoder, SequentialEncoder
 
-__all__ = ["transformed_image_loss"]
+from ._modules import (
+    DiscriminatorMultiLayerEncoder,
+    TransformerBlock,
+    prediction_module,
+)
+
+__all__ = [
+    "EncodingDiscriminatorOperator",
+    "PredictionOperator",
+    "MultiLayerPredictionOperator",
+    "prediction_loss",
+    "DiscriminatorLoss",
+    "discriminator_loss",
+    "transformed_image_loss",
+]
+
+
+class EncodingDiscriminatorOperator(ops.EncodingRegularizationOperator):
+    r"""Abstract base class for all discriminator operators working in an encoded space.
+
+    In this operator a discriminator mode can be set, in whose dependence the output can
+    be influenced. This can be either ``real`` or ``fake``. In addition, it is
+    calculated how accurate the operator was on the last input.
+
+    Args:
+        encoder: Encoder that is used to encode the target and input images.
+        score_weight: Score weight of the operator. Defaults to ``1.0``.
+    """
+
+    def __init__(self, encoder: Encoder, score_weight: float = 1.0):
+        super().__init__(encoder, score_weight=score_weight)
+        self._target_distribution = True
+        self.accuracy: torch.Tensor
+        self.register_buffer("accuracy", torch.zeros(1))
+
+    def real(self, mode: bool = True) -> "EncodingDiscriminatorOperator":
+        self._target_distribution = mode
+        return self
+
+    def fake(self) -> "EncodingDiscriminatorOperator":
+        return self.real(False)
+
+    def process_input_image(self, image: torch.Tensor) -> torch.Tensor:
+        input_repr = self.input_image_to_repr(image)
+        self.accuracy = self.calculate_accuracy(input_repr)
+        return self.calculate_score(input_repr)
+
+    @abstractmethod
+    def calculate_accuracy(self, input_repr: torch.Tensor) -> torch.Tensor:
+        pass
+
+
+class PredictionOperator(EncodingDiscriminatorOperator):
+    r"""The prediction loss is a discriminator loss based on the prediction.
+
+    The prediction consists of the output of an ``predictor`` which processes the
+    output of the ``encoder``.
+
+    It measures the cross-entropy loss between true labels and predicted labels. The
+    true labels are set depending on the currently set discriminator mode.
```

Comment:
This should be clear from the docstring of the `EncodingDiscriminatorOperator` or any of its methods. If it is not, please fix it there.

```suggestion
    It measures the cross-entropy loss between true labels and predicted labels. 
```